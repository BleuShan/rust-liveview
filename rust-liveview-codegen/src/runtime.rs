use crate::helpers::{
    has_attr,
    DarlingResultExt,
};
use darling::FromMeta;
use proc_macro::TokenStream;
use proc_macro_error::*;
use quote::quote;
use syn::{
    AttributeArgs,
    Error,
    ItemFn,
    Path,
};

#[derive(Debug, FromMeta)]
pub(crate) struct Attribute {
    executor_entrypoint: Path,
    #[darling(default)]
    test: bool,
}

impl From<AttributeArgs> for Attribute {
    fn from(attr_args: AttributeArgs) -> Self {
        Self::from_list(&attr_args).expect_or_abort("An error occured while parsing the input.")
    }
}

impl Attribute {
    pub(crate) fn generate(self, mut item: ItemFn) -> TokenStream {
        let sig = &mut item.sig;

        if sig.asyncness.is_none() {
            abort!(Error::new_spanned(
                sig.fn_token,
                "Only async functions are supported"
            ));
        }

        let vis = &item.vis;
        let attrs = &item.attrs;
        let body = &item.block;
        let entrypoint = &self.executor_entrypoint;

        sig.asyncness = None;

        let attributes = if self.test && !has_attr(attrs, "test") {
            quote! {
                #[test]
                #(#attrs)*
            }
        } else {
            quote!(#(#attrs)*)
        };
        TokenStream::from(quote! {
            #attributes
            #vis #sig {
                #entrypoint(async move {
                    #body
                })
            }
        })
    }
}
