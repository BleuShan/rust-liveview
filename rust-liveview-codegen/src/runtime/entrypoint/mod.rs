use crate::helpers::{
    new,
    DarlingResultExt,
    Display,
};
use darling::FromMeta;
use proc_macro::{
    Span,
    TokenStream,
};
use proc_macro2::TokenStream as TokenStream2;
use proc_macro_error::*;
use quote::{
    quote,
    ToTokens,
};
use syn::{
    parse::{
        Parse,
        ParseStream,
        Result,
    },
    punctuated::{
        Pair,
        Punctuated,
    },
    spanned::Spanned,
    AttributeArgs,
    FnArg,
    ItemFn,
    NestedMeta,
    Token,
};

mod main;
mod test;

pub(crate) use main::{
    MainEntryPoint,
    MainEntryPointArgs,
};
pub(crate) use test::{
    TestEntryPoint,
    TestEntryPointArgs,
};

#[inline]
pub(crate) fn set_fn_dummy(item: &ItemFn) {
    let sig = &item.sig;
    let attrs = &item.attrs;
    let ident = &sig.ident;
    let inputs = &sig.inputs;
    let vis = &item.vis;
    set_dummy(quote! {
        #(#attrs)*
        #vis fn #ident(#inputs) {}
    });
}

#[derive(Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, FromMeta, Debug, Display)]
pub(crate) enum Executor {
    AsyncStd,
    Tokio,
}
