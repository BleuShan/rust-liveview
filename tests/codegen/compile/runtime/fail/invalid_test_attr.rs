#[allow(unused_imports)]
use async_std::task;
use rust_liveview::runtime;

fn main() {}

#[runtime::test(executor = "async_std")]
#[test]
async fn runtime_attribute_test_should_add_the_test_attribute() {}
