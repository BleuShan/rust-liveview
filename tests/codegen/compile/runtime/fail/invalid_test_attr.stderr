error: Second test attribute is supplied.
 --> $DIR/invalid_test_attr.rs:8:1
  |
8 | #[test]
  | ^^^^^^^
  |
  = note: test attribute is supplied by default
